### PR TITLE
Feature/update requirements

### PR DESCRIPTION
**Pull request by wern**
As discussed we adapted the expected results of some of the tests and move tests dealing with not mandatory feature into to optional section.

What was changed in detail:

These should test for "valid answer" not for success (eg. HTTP-400 is a valid answer, eg. "redirect" to asyc API):
Test Case 003: Get All Footprints
Test Case 004: Get Limited List of Footprints
Test Case 008: Get Footprint
Test Case 011: Attempt GetFootprint with Non-Existent PfId

These tests should be optional as the feature they test is optional:
Test Case 012: Receive Notification of PCF Update
Test Case 013: Notify of PCF Update

In addition we had an internal discussion regarding the definiition of an "invalid token". This should be described in more detail (a "broken" token missing half of the data, a token with the wrong scope, ...). As based on the definition we should expect an HTTP-403 (Forbidden) or a HTTP-401 (Unauthorized) - see also: https://www.permit.io/blog/401-vs-403-error-whats-the-difference . This should be defined clearer in another PR.

### CODE REVIEW

**Comment by wern** [comment]
@ArunavC as I cannot assign reviewers, I loop you in here this way. ;) This is the PR we wanted to raise up front to the next working sesssion, so it can be presented there and maybe finds some reviewers.

File: checklist.bs [hunk 1]
```diff
@@ -157,19 +168,27 @@ system with the **limit** parameter, a **valid access token** and the syntax spe
...
+- 200 OK and a JSON body containing the list of all
+    available PCFs, as specified in [[DXP#api-action-list-response]]
+    Unless the total number of available PCFs is equal to or smaller 
+    than the limit set in the request, the test target host system 
+    must return a `Link` header, or
+
+- 202 ACCEPTED and a JSON body containing the list of some
+    available PCFs, as specified in [[DXP#api-action-list-response]], or
+
+- 400 BAD REQUEST and a JSON body containing an error, as specified in
```
Comment (zeitgeist) [inline checklist.bs:180]: Acccording to this test definition, **always** returning a 400 (with a `BadRequest` for instance) was conforming behavior. 

Is this intentionally so?

File: checklist.bs [hunk 2]
```diff
@@ -239,6 +258,15 @@ host system with a **valid access token** and the syntax specified in
 The test target host system must respond with 200 OK and a JSON body containing the PCF with the
 requested `pfId`, as specified in [[DXP#api-action-get-response]].
 
+The test target host system must respond with
+- 200 OK and a JSON body containing the PCF with the
+    requested `pfId`, as specified in [[DXP#api-action-get-response]], or
+
+- 400 BAD REQUEST and a JSON body containing an error, as specified in
```
Comment (zeitgeist) [inline checklist.bs:265]: Likewise (see prev. comment for Test Case 004)

File: checklist.bs [hunk 3]
```diff
@@ -289,37 +317,41 @@ syntax specified in [[DXP#api-action-get-request]].
 The test target host system should respond with a 404 Not Found and a JSON body containing the error
 code `NoSuchFootprint`, as specified in [[DXP#api-error-responses]].
 
-## Test Case 012: Receive Notification of PCF Update ## {#tc012}
+The test target host system must respond with
+- 404 NOT FOUND and a JSON body containing the error code `NoSuchFootprint`,
+    as specified in [[DXP#api-error-responses]], or
+
+- 400 BAD REQUEST and a JSON body containing an error, as specified in
```
Comment (zeitgeist) [inline checklist.bs:324]: Likewise, always returning a 400 would be conformant behavior, effectively turning the behavior and this endpoint in general into an "OPTIONAL" one.

**Review by zeitgeist — COMMENTED** [review]
Thanks for the many updates and corrections Werner. One remark / concern find shared across the document.

File: checklist.bs (hunk 1 above)
Comment (wern) [reply checklist.bs:180]: Yes, it is as this is backed up by the specification. We did this change to the specs to ease implementation of the standard for solutions that are primarily based on asynchronous communication. Although target is to also return PCFs in a synchronous way, the missing ability to do it in first shot should not prevent a solution from being compliant (as this is defined in the specs as a legal answer). As long as the endpoints are provided and operational as described, of course.

File: checklist.bs (hunk 3 above)
Comment (wern) [reply checklist.bs:324]: If the endpoint would be "Optional" a solution would not have to implement it. This would break existing clients. In the spec changes we introduces the "Http-400 => go to async" approach to explicitly avoid this situation. Endpoint needs to be there and to return valid answers (in respect to the specification).

File: checklist.bs (hunk 1 above)
Comment (zeitgeist) [reply checklist.bs:180]: I think the difference is the understanding of "operational" and what the tech specs aim to bring forward.

My personal $0.02 and understanding of the tech specs are:

a) the endpoint does not return an error unless something is really wrong (i.e. authentication failed). If there is nothing there yet to be shared, the empty list is returned.
b) if the data sought for is not there, the protocol to request the PCFs is followed
c) after this succeeded and PCF data is now there, the endpoint again does not return an error. However, it does return PCFs that were successfully requested.

The way I read the diff here, this is no longer the case. You can always return an error and ask everyone to go through the async processing.

Do you agree with my understanding of your proposal, @wern , or am I missing something?

Comment (schuur) [reply checklist.bs:180]: Reading through the specification, there's no explicit way of signalling that asynchronous processing must be used to get footprints after an attempt to get them via a simple synchronous listfootprints. 

I mistakenly thought at first that a HTTP status code **202** could be used for that precise purpose, but looking at the specs, a **202** only invites repeated synchronous calls until a **200** indicates completeness.

Given these constraints, a **400** would be a very blunt instrument to redirect the client to the async method of getting PCFs. 

Given the current specification, there's no other way to indicate this, so let's accept it.

For version 3 however, I would rather propose a change in using **202 (Accepted)** as a way of indicating asynchronous processing and **206 (Partial result**) as an invitation to retry synchronously.

File: checklist.bs (hunk 2 above)
Comment (schuur) [reply checklist.bs:265]: And see above as well ;)

File: checklist.bs (hunk 3 above)
Comment (schuur) [reply checklist.bs:324]: In the absence of 2xx codes for signalling async processing, using 400 for this one is in line with list footprints. 

Also for this one, a 202 (Accepted) would be a good solution, I think.

File: checklist.bs (hunk 1 above)
Comment (zeitgeist) [reply checklist.bs:180]: Thanks for your commend, @schuur . Replying here as I am listed as a reviewer.

> Reading through the specification, there's no explicit way of signalling that asynchronous processing must be used to get footprints after an attempt to get them via a simple synchronous listfootprints.

That is correct.  However, the tech specs already have full support for this:

1. You call `ListFootprints`
2. If the footprint you are looking for is missing, you follow this guidance here: https://wbcsd.github.io/data-exchange-protocol/v2/#business-cases-async-events-1

TL;DR – for my reading of the situation, the `400` is not aligned with the tech specs (and the assumptions implementers made), but more importantly, is not needed as a signalling feature at all. The support is already there and it is document.

If you do not agree with this reading, and more explicit signalling was indeed needed, I would argue the conformance testing documentation is unfortunately **not** the right place for making a decision like this. 

The tech specs should then be updated accordingly, undergoing a feedback process with the other SPs especially.

Comment (wern) [reply checklist.bs:180]: @zeitgeist We were both part of the discussion to use HTTP-303 to signal this. This was postponed to version 3 as it is supposed to be a breaking change. So we are not opening a new discussion here, but continuing the old one we already had in the tech spec group. This PR tries to build a bridge until the new spec comes with a corresponding return code (best suggestion made until now: HTTP-303), as we discussed that implicit logic like "you don't find it here, look over there" is not supposed to be good API style. So in V2.2 HTTP-400 is the best we have, from my point of view, to make things explicit. 

If now a majority agrees here that returning an empty (or incomplete) list is preferred (what would revert the result of the discussion done for the tech spec!), we should put this into the definition of the tests. But right now they are too strict and need to be adapted.

**Review by ArunavC — COMMENTED** [review]
Looks good
